### PR TITLE
Feature/fga/reactions UI improvements

### PR DESCRIPTION
**Pull request by ganfra**
This PR improves the UI of reactions by handling #3344 #4674 and #5149
It also fixes reactions order on Bubble and UI echo.

| COLLAPSED |EXPANDED | 
| --- | --- |
Screenshot_20220210-190439|Screenshot_20220210-190444

### CODE REVIEW

**Comment by amshakal** [comment]
Looks good! 

The designs were made with the assumption that the text in each reaction is black in colour. Can we match the colour and style of '3 more' and 'show less' to match the text style used for reactions? i.e. grey and bold text?

**Comment by ganfra** [comment]
> will this also fix #5149 ?

Oh yeah probably!

File: vector/src/main/java/im/vector/app/features/home/room/detail/timeline/item/AbsBaseMessageItem.kt [hunk 1]
```diff
@@ -83,23 +113,36 @@ abstract class AbsBaseMessageItem<H : AbsBaseMessageItem.Holder> : BaseEventItem
                 reactionButton.isEnabled = reaction.synced
                 holder.reactionsContainer.addView(reactionButton)
             }
+            if (reactions.count() > MAX_REACTIONS_TO_SHOW) {
+                val showReactionsTextView = createReactionTextView(holder, 6)
```
Comment (ouchadam) [inline vector/src/main/java/im/vector/app/features/home/room/detail/timeline/item/AbsBaseMessageItem.kt:117]: could be handy to have some named arguments for the hardcoded values for some extra context

**Review by ouchadam — APPROVED** [review]
LGTM as well! :100:

File: vector/src/main/java/im/vector/app/features/home/room/detail/timeline/item/AbsBaseMessageItem.kt (hunk 1 above)
Comment (ganfra) [reply vector/src/main/java/im/vector/app/features/home/room/detail/timeline/item/AbsBaseMessageItem.kt:117]: Actually, I'm changing this to a Style as it's shared with ReactionButton

**Review by amshakal — APPROVED** [review]
TY!

File: vector/src/main/java/im/vector/app/features/home/room/detail/timeline/item/MessageInformationData.kt [hunk 2]
```diff
@@ -55,6 +54,16 @@ data class ReferencesInfoData(
...
 ) : Parcelable
 
+@Parcelize
+data class ReactionsSummaryData(
+        /*List of reactions (emoji,count,isSelected)*/
+        val reactions: List<ReactionInfoData>? = null,
+        val showAll: Boolean = false,
+        val onShowMoreClicked: () -> Unit,
+        val onShowLessClicked: () -> Unit,
+        val onAddMoreClicked: () -> Unit
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/home/room/detail/timeline/item/MessageInformationData.kt:64]: Weird to see some lambda in data class

File: vector/src/main/res/values/strings.xml [hunk 3]
```diff
@@ -3779,4 +3779,7 @@
     <string name="tooltip_attachment_poll">Create poll</string>
     <string name="tooltip_attachment_location">Share location</string>
 
+    <string name="message_reaction_show_less">Show less</string>
+    <string name="message_reaction_show_more">"%1$d more"</string>
```
Comment (bmarty) [inline vector/src/main/res/values/strings.xml:3783]: You should have used a `plurals` :/

File: vector/src/main/java/im/vector/app/features/home/room/detail/timeline/item/MessageInformationData.kt (hunk 2 above)
Comment (ouchadam) [reply vector/src/main/java/im/vector/app/features/home/room/detail/timeline/item/MessageInformationData.kt:64]: not sure what will happen when we attempt to parcelize :thinking:

**Pull request merged**